src/gui/torrentcontentmodel.cpp

```diff
@@ -319,8 +323,45 @@ void TorrentContentModel::setupModelData(const BitTorrent::TorrentInfo &info)
         TorrentContentModelFile* fileItem = new TorrentContentModelFile(info.fileName(i), info.fileSize(i), currentParent, i);
         currentParent->appendChild(fileItem);
         m_filesIndex.push_back(fileItem);
+
+        if (!Preferences::instance()->getAutoFilePriorities() || TorrentContentModelItem::FileType != fileItem->itemType())
+            continue;
+        else
+            prioritySortList.append({path, fileItem});
     }
     emit layoutChanged();
+
+    if (prioritySortList.size() < 2) return;
+
+    // increase priorities for the first files in the list so these are dowloaded first
+    qSort(prioritySortList.begin(), prioritySortList.end(), [&prioType](const Entry& e1, const Entry& e2) {
+        using AutoPrio = BitTorrent::FileAutoPriority;
+        switch (prioType) {
+        case AutoPrio::NAME_DESC:
+            return e1.first.compare(e2.first, Qt::CaseInsensitive) > 0;
+        case AutoPrio::SIZE_ASC:
+            return e1.second->size() < e2.second->size();
+        case AutoPrio::SIZE_DESC:
+            return e1.second->size() > e2.second->size();
+        default:
+            return e1.first.compare(e2.first, Qt::CaseInsensitive) < 0; // NAME_ASC
+        }
+    });
+
+    QVector<int> priorities;
+    switch (prioritySortList.size()) {
+    case 2:
+        priorities.push_back(prio::HIGH);
+        break;
+    case 3:
+        priorities = QVector<int>({prio::MAXIMUM, prio::HIGH});
+        break;
+    default:
+        priorities = QVector<int>({prio::MAXIMUM, prio::HIGH, prio::HIGH});
+    }
+
+    for (int i = 0; i < priorities.size(); ++i)
+        prioritySortList[i].second->setPriority(priorities[i]);
```

Comment:
The second BLOCKING issue - TorrentContentModel (i.e. Presentation Model) is WRONG place for such a logic.